aiida/backends/sqlalchemy/migrations/versions/d102e96ea943_inline_calc_attributes.py

```diff
@@ -0,0 +1,130 @@
+the relevant code here (line 60 has the execution call, it is stored after that):
+
+https://github.com/aiidateam/aiida-core/blob/v0.12.5/aiida/orm/implementation/django/calculation/inline.py#L60
+
+Revision ID: d102e96ea943
+Revises: 0edcdd5a30f0
+Create Date: 2021-02-09 16:25:16.106176
+
+"""
+# pylint: disable=invalid-name,no-member
+# yapf: disable
+from alembic import op
+from sqlalchemy.sql import text
+from aiida.backends.general.migrations import utils
+
+# revision identifiers, used by Alembic.
+revision = 'd102e96ea943'
+down_revision = '0edcdd5a30f0'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    """Creates the new attributes"""
+    conn = op.get_bind()
+
+    # Creates 'function_namespace' from 'namespace'
+    # Creates 'function_starting_line_number' from 'first_line_source_code'
+    # Creates 'exit_status', 'process_state' and 'process_label'
+    conn.execute(text(
+        """
+        UPDATE db_dbnode
+        SET attributes = attributes || '{
+            "exit_status": 0,
+            "process_state": "finished",
+            "process_label": "Legacy InlineCalculation"
+            }'::jsonb ||
+            jsonb_build_object(
+                'function_starting_line_number', attributes->'first_line_source_code',
+                'function_namespace', attributes->'namespace'
+                )
+        WHERE
+            attributes ? 'namespace' AND
+            attributes ? 'first_line_source_code' AND
+            node_type = 'process.calculation.calcfunction.CalcFunctionNode.';
+        """))
+
+
+    # Sets new 'function_number_of_lines' with the number of lines in 'source_code'
+    # What counts as a newline (SOURCE https://stackoverflow.com/a/10805292)
+    # Why is it done as a difference + division (https://stackoverflow.com/a/36376548)
+    replace_dicts = [
+        {'symbol': '\\r\\n', 'symlike': '%\\r\\n%', 'count': 6},  # WinDOS
+        {'symbol': '\\r', 'symlike': '%\\r%', 'count': 3},  # older Macs
+        {'symbol': '\\n', 'symlike': '%\\n%', 'count': 3},  # Unices
+    ]
+
+    for replace_dict in replace_dicts:
+        conn.execute(text(
+            """
+            UPDATE db_dbnode
+            SET attributes = attributes || jsonb_build_object(
+                'function_number_of_lines',
+                ( CHAR_LENGTH((attributes -> 'source_code')::TEXT)
+                - CHAR_LENGTH( REPLACE((attributes -> 'source_code')::TEXT, :symbol, '' )))
+                / :count + 1
+                )
+            WHERE
+                attributes ? 'namespace' AND
+                attributes ? 'first_line_source_code' AND
+                node_type = 'process.calculation.calcfunction.CalcFunctionNode.' AND
+                (attributes -> 'source_code')::TEXT LIKE :symlike;
+            """), replace_dict)
+
+
+    # Saves the source file in the repository
+    sqlquery_results = conn.execute(text("SELECT uuid, attributes -> 'source_file' FROM db_dbnode;"))
```

Comment:
shouldn't we put another filter to make sure we only target legacy migrated calc nodes to be sure? Here you are fetching all nodes with the attribute `source_file`. There could be others

Reply:
Good point. I modified the `WHERE` clause of all other queries to be even more specific and added it here as well:

```SQL
        SELECT uuid, attributes -> 'source_file'
        FROM db_dbnode
        WHERE
            attributes ? 'namespace' AND
            attributes ? 'source_code' AND
            attributes ? 'source_file' AND
            attributes ? 'first_line_source_code' AND
            node_type = 'process.calculation.calcfunction.CalcFunctionNode.';
```